README.md
random5
=======

A simple XBMC/Kodi add-on that plays 5 (or how ever many you define in the add-on's settings) random episodes from a 
chosen TV show. Easy way to decide what to watch when falling asleep.

### Settings ###
* Number of playlist items - Determines the amount of episodes added to the playlist.
* Watched episodes only - Determines if only watched items should be added to the playlist.
* Sequential order - If enabled the first episode is randomly chosen and the others follow in order.
* Automatically start playback - If enabled playback of the playlist is automatically started otherwise the playlist
is simply shown.
* Show 'Random Show' - Show the 'Random Show' item for times of indecisiveness. 

changelog.txt
v0.4.1 ( 09 March 2016)
- Account for shows not having all types of artwork

v0.4.0 ( 03 August 2015 )
- Sort shows by title (ascending)

v0.3.2
- Graceful exit when no TV shows found in library

v0.3.1
- Really no more hardcoded strings!

v0.3.0
- Artwork for the different views
- No more hardcoded strings

default.py
import sys
import random
import urlparse
import json
from StringIO import StringIO

import xbmc
import xbmcgui
import xbmcplugin
import xbmcaddon


addon_handle = int(sys.argv[1])

__addon__ = xbmcaddon.Addon(id='plugin.video.random5')
__addonname__ = __addon__.getAddonInfo('name')

# Get settings
number = int(__addon__.getSetting('number_of_episodes'))
watched_only = __addon__.getSetting('watched_only')
sequential_order = __addon__.getSetting('sequential_order')
autoplay = __addon__.getSetting('autoplay')
random_show_list_item = __addon__.getSetting('random_show_list_item')

args = urlparse.parse_qs(sys.argv[2][1:])
path = args.get('path', None)


def log(text, level):
    if isinstance(text, str):
        text = text.decode('utf-8')

    message = u'{0}: {1}'.format(__addonname__, text)
    xbmc.log(msg=message.encode('utf-8'), level=level)


def execute_json(method, *args, **kwargs):
    if len(args) == 1:
        args = args[0]
    else:
        args = kwargs

    params = {'jsonrpc': '2.0', 'id': 0, 'method': method, 'params': args}

    values = json.dumps(params)

    json_command = values
    json_result = xbmc.executeJSONRPC(json_command)

    return json.load(StringIO(json_result))


def get_shows():
    """Returns all shows in the library"""
    request = execute_json('VideoLibrary.GetTVShows', { "sort": {"method": "tvshowtitle"} })

    try:
        shows = request['result']['tvshows']
        log('Returning %s shows' % len(shows), level=xbmc.LOGDEBUG)
        return shows
    except KeyError:
        xbmcgui.Dialog().ok(__addonname__, __addon__.getLocalizedString(30100))
        log('No shows found in library! Exiting...', level=xbmc.LOGDEBUG)
        sys.exit(-1)

def get_show_details(show, properties):
    request = execute_json('VideoLibrary.GetTVShowDetails', {"tvshowid": show, "properties": properties})
    return request['result']['tvshowdetails']
    

def get_episodes_by_show(show):
    """Gets all episodes for given show"""
    show = int(show)  # Needs to be an integer

    request = execute_json('VideoLibrary.GetEpisodes', {'tvshowid': show, 'properties': ['title', 'file', 'playcount']})
    episodes = request['result']['episodes']

    if watched_only == 'true':
        episodes = get_watched_episodes(episodes)

    log('Returning {0} episodes to start with'.format(len(episodes)), level=xbmc.LOGDEBUG)
    return episodes


def add_file_to_playlist(file):
    """Adds a video file to the video playlist"""
    execute_json('Playlist.Add', {"item": {"file": file['file']}, "playlistid": 1})
    log('Adding {0} to playlist'.format(file['label']), level=xbmc.LOGDEBUG)


def get_watched_episodes(episode_list):
    """Takes a list of episodes and returns only episodes that have a playcount of > 0"""
    watched_episodes = []
    for episode in episode_list:
        if episode['playcount'] > 0:
            watched_episodes.append(episode)

    return watched_episodes


def get_episodes():
    """Returns all episodes"""
    request = execute_json('VideoLibrary.GetEpisodes', {'properties': ['title', 'tvshowid', 'file', 'playcount']})
    episodes = request['result']['episodes']

    if watched_only == 'true':
        episodes = get_watched_episodes(episodes)

    return episodes


def get_sequential_episodes(number, show):
    episodes = []
    show_episodes = get_episodes_by_show(show)

    if len(show_episodes) == 0:
        line1 = __addon__.getLocalizedString(30100)
        line2 = __addon__.getLocalizedString(30101)
        xbmcgui.Dialog().ok(__addonname__, line1, line2)

    if len(show_episodes) < number:  # Just start with the first episode
        first_episode = show_episodes[0]
        episodes.append(first_episode)

    else:  # Get a random episode from the show and add it to the list
        first_episode = random.choice(show_episodes)
        episodes.append(first_episode)

    # Next episode is the index of the first episode + 1
    next_episode = show_episodes.index(first_episode) + 1

    while len(episodes) < number:
        try:
            episode = show_episodes[next_episode]
            episodes.append(episode)
            next_episode += 1
        except IndexError:
            break

    return episodes


def get_random_episodes(number, show):
    """Randomly selects a number of episodes from a particular show"""
    episodes = []
    show_episodes = get_episodes_by_show(show)

    # Check to make sure we still have an episode pool to pull from
    if len(show_episodes) < 1:
        line1 = __addon__.getLocalizedString(30100)
        line2 = __addon__.getLocalizedString(30101)
        xbmcgui.Dialog().ok(__addonname__, line1, line2)

    if number <= len(show_episodes):
        while len(episodes) < number:
            random_episode = random.choice(show_episodes)
            if random_episode not in episodes:
                episodes.append(random_episode)
                log('Randomly chose episode {0} with playcount {1}'.format(
                    random_episode['label'], random_episode['playcount']), level=xbmc.LOGDEBUG)

    # If there are not enough episodes to meet the desired number of playlist items, add what episodes there are
    elif number > len(show_episodes):
        while len(episodes) < len(show_episodes):
            random_episode = random.choice(show_episodes)
            if random_episode not in episodes:
                episodes.append(random_episode)
                log('Randomly chose episode {0} with playcount {1}'.format(
                    random_episode['label'], random_episode['playcount']), level=xbmc.LOGDEBUG)

    return episodes


def create_playlist(episode_list):
    """Creates a playlist of episodes of a particular show"""
    # Clear the playlist before we add items to it
    execute_json('Playlist.Clear', {"playlistid": 1})

    for item in episode_list:
        add_file_to_playlist(item)

    log('Playlist created', level=xbmc.LOGDEBUG)


def create_and_play(show):
    """Creates playlist and plays playlist"""
    if sequential_order == 'true':
        episodes = get_sequential_episodes(number, show)
    else:
        episodes = get_random_episodes(number, show)

    # Don't bother doing any of the below if there are no episodes
    if len(episodes) > 0:
        create_playlist(episodes)

        if autoplay == 'true':
            execute_json('Player.Open', {"item": {"playlistid": xbmc.PLAYLIST_VIDEO}})
            log('Playing playlist', level=xbmc.LOGDEBUG)
        else:
            execute_json('GUI.ActivateWindow', {'window': 'videoplaylist'})
            log('Showing playlist', level=xbmc.LOGDEBUG)


def create_menu(show_list):
    """Creates the 'folder' menu"""

    # TODO: Create 'entire library' list item
    # li = xbmcgui.ListItem('ALL TV SHOWS', iconImage='DefaultFolder.png')
    # xbmcplugin.addDirectoryItem(handle=addon_handle, url='show', listitem=li, isFolder=False)

    # 'Random show' entry
    if random_show_list_item == 'true':
        random_show = random.choice(show_list)
        url = sys.argv[0] + "?path=/play&show=%s" % (random_show['tvshowid'])
        li = xbmcgui.ListItem('- RANDOM SHOW -', iconImage='DefaultFolder.png')
        xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=li, isFolder=False)

    for s in show_list:
        url = sys.argv[0] + "?path=/play&show=%s" % (s['tvshowid'])
        show_artwork = get_show_details(s['tvshowid'], ['art'])['art']
        
        li = xbmcgui.ListItem(s['label'], iconImage=show_artwork['poster'], thumbnailImage=show_artwork['poster'])
        li.setArt({'banner': show_artwork['banner'], 'fanart': show_artwork['fanart']})
        xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=li, isFolder=False)
    
    xbmcplugin.endOfDirectory(addon_handle)

# The 'navigation' part
if path is None:
    create_menu(get_shows())

elif path[0] == '/play':
    showid = args.get('show', None)
    create_and_play(showid[0])

resources/lib/xbmcjson/xbmcjson.py
#!/bin/env python

import urllib, urllib2
import json
from StringIO import StringIO

PLAYER_VIDEO=1

class XBMCTransport(object):
  """Base class for XBMC transport"""
  def execute(self, method, args):
    pass

class XBMCJsonTransport(XBMCTransport):
  """HTTP Json transport"""
  def __init__(self, url, username='xbmc', password='xbmc'):
    self.url=url
    self.username=username
    self.password=password
    self.id = 0

  def execute(self, method, *args, **kwargs):
    header = {
        'Content-Type' : 'application/json',
        'User-Agent' : 'python-xbmc'
        }
    # Params are given as a dictionnary
    if len(args) == 1:
      args=args[0]
    # Use kwargs for param=value style
    else:
      args = kwargs
    params={}
    params['jsonrpc']='2.0'
    params['id']=self.id
    self.id +=1
    params['method']=method
    params['params']=args

    values=json.dumps(params)
    # HTTP Authentication
    password_mgr = urllib2.HTTPPasswordMgrWithDefaultRealm() 
    password_mgr.add_password(None, self.url, self.username, self.password) 
    auth_handler = urllib2.HTTPBasicAuthHandler(password_mgr)
    opener = urllib2.build_opener(auth_handler) 
    urllib2.install_opener(opener)

    data = values
    req = urllib2.Request(self.url, data, header)
    response = urllib2.urlopen(req)
    the_page = response.read()
    print the_page
    if len(the_page) > 0 :
      return json.load(StringIO(the_page))
    else:
      return None # for readability

class XBMC(object):
  """XBMC client"""
  def __init__(self, url, username='xbmc', password='xbmc'):
    self.transport = XBMCJsonTransport(url, username, password)
    # Dynamic namespace class instanciation
    for cl in namespaces:
      s = "self.%s = %s(self.transport)"%(cl,cl)
      exec(s)
    def execute(self, *args, **kwargs):
      self.transport.execute(*args, **kwargs)

class XBMCNamespace(object):
  """Base class for XBMC namespace."""
  def __init__(self, xbmc):
    self.xbmc = xbmc
  def __getattr__(self, name):
    klass= self.__class__.__name__
    method=name
    xbmcmethod = "%s.%s"%(klass, method)
    def hook(*args, **kwargs):
      return self.xbmc.execute(xbmcmethod, *args, **kwargs)
    return hook

# Dynamic namespace class injection
namespaces = ["VideoLibrary", "AudioLibrary", "Application", "Player", "Input", "System", "Playlist", "Addons", "AudioLibrary", "Files", "GUI" , "JSONRPC", "PVR", "xbmc"]
for cl in namespaces:
  s = """class %s(XBMCNamespace):
  \"\"\"XBMC %s namespace. \"\"\"
  pass
  """%(cl,cl)
  exec (s)


addon.py
import sys
import random
import urlparse
import json
from StringIO import StringIO

import xbmc
import xbmcgui
import xbmcplugin
import xbmcaddon


addon_handle = int(sys.argv[1])

__addon__ = xbmcaddon.Addon(id='plugin.video.random5')
__addonname__ = __addon__.getAddonInfo('name')

# Get settings
number = int(__addon__.getSetting('number_of_episodes'))
watched_only = __addon__.getSetting('watched_only')
sequential_order = __addon__.getSetting('sequential_order')
autoplay = __addon__.getSetting('autoplay')
random_show_list_item = __addon__.getSetting('random_show_list_item')

args = urlparse.parse_qs(sys.argv[2][1:])
path = args.get('path', None)


def log(text, level):
    if isinstance(text, str):
        text = text.decode('utf-8')

    message = u'{0}: {1}'.format(__addonname__, text)
    xbmc.log(msg=message.encode('utf-8'), level=level)


def execute_json(method, *args, **kwargs):
    if len(args) == 1:
        args = args[0]
    else:
        args = kwargs

    params = {'jsonrpc': '2.0', 'id': 0, 'method': method, 'params': args}

    values = json.dumps(params)

    json_command = values
    json_result = xbmc.executeJSONRPC(json_command)

    return json.load(StringIO(json_result))


def get_shows():
    """Returns all shows in the library"""
    request = execute_json('VideoLibrary.GetTVShows', { "sort": {"method": "tvshowtitle"} })

    try:
        shows = request['result']['tvshows']
        log('Returning %s shows' % len(shows), level=xbmc.LOGDEBUG)
        return shows
    except KeyError:
        xbmcgui.Dialog().ok(__addonname__, __addon__.getLocalizedString(30100))
        log('No shows found in library! Exiting...', level=xbmc.LOGDEBUG)
        sys.exit(-1)

def get_show_details(show, properties):
    request = execute_json('VideoLibrary.GetTVShowDetails', {"tvshowid": show, "properties": properties})
    return request['result']['tvshowdetails']
    

def get_episodes_by_show(show):
    """Gets all episodes for given show"""
    show = int(show)  # Needs to be an integer

    request = execute_json('VideoLibrary.GetEpisodes', {'tvshowid': show, 'properties': ['title', 'file', 'playcount']})
    episodes = request['result']['episodes']

    if watched_only == 'true':
        episodes = get_watched_episodes(episodes)

    log('Returning {0} episodes to start with'.format(len(episodes)), level=xbmc.LOGDEBUG)
    return episodes


def add_file_to_playlist(file):
    """Adds a video file to the video playlist"""
    execute_json('Playlist.Add', {"item": {"file": file['file']}, "playlistid": 1})
    log('Adding {0} to playlist'.format(file['label']), level=xbmc.LOGDEBUG)


def get_watched_episodes(episode_list):
    """Takes a list of episodes and returns only episodes that have a playcount of > 0"""
    watched_episodes = []
    for episode in episode_list:
        if episode['playcount'] > 0:
            watched_episodes.append(episode)

    return watched_episodes


def get_episodes():
    """Returns all episodes"""
    request = execute_json('VideoLibrary.GetEpisodes', {'properties': ['title', 'tvshowid', 'file', 'playcount']})
    episodes = request['result']['episodes']

    if watched_only == 'true':
        episodes = get_watched_episodes(episodes)

    return episodes


def get_sequential_episodes(number, show):
    episodes = []
    show_episodes = get_episodes_by_show(show)

    if len(show_episodes) == 0:
        line1 = __addon__.getLocalizedString(30100)
        line2 = __addon__.getLocalizedString(30101)
        xbmcgui.Dialog().ok(__addonname__, line1, line2)

    if len(show_episodes) < number:  # Just start with the first episode
        first_episode = show_episodes[0]
        episodes.append(first_episode)

    else:  # Get a random episode from the show and add it to the list
        first_episode = random.choice(show_episodes)
        episodes.append(first_episode)

    # Next episode is the index of the first episode + 1
    next_episode = show_episodes.index(first_episode) + 1

    while len(episodes) < number:
        try:
            episode = show_episodes[next_episode]
            episodes.append(episode)
            next_episode += 1
        except IndexError:
            break

    return episodes


def get_random_episodes(number, show):
    """Randomly selects a number of episodes from a particular show"""
    episodes = []
    show_episodes = get_episodes_by_show(show)

    # Check to make sure we still have an episode pool to pull from
    if len(show_episodes) < 1:
        line1 = __addon__.getLocalizedString(30100)
        line2 = __addon__.getLocalizedString(30101)
        xbmcgui.Dialog().ok(__addonname__, line1, line2)

    if number <= len(show_episodes):
        while len(episodes) < number:
            random_episode = random.choice(show_episodes)
            if random_episode not in episodes:
                episodes.append(random_episode)
                log('Randomly chose episode {0} with playcount {1}'.format(
                    random_episode['label'], random_episode['playcount']), level=xbmc.LOGDEBUG)

    # If there are not enough episodes to meet the desired number of playlist items, add what episodes there are
    elif number > len(show_episodes):
        while len(episodes) < len(show_episodes):
            random_episode = random.choice(show_episodes)
            if random_episode not in episodes:
                episodes.append(random_episode)
                log('Randomly chose episode {0} with playcount {1}'.format(
                    random_episode['label'], random_episode['playcount']), level=xbmc.LOGDEBUG)

    return episodes


def create_playlist(episode_list):
    """Creates a playlist of episodes of a particular show"""
    # Clear the playlist before we add items to it
    execute_json('Playlist.Clear', {"playlistid": 1})

    for item in episode_list:
        add_file_to_playlist(item)

    log('Playlist created', level=xbmc.LOGDEBUG)


def create_and_play(show):
    """Creates playlist and plays playlist"""
    if sequential_order == 'true':
        episodes = get_sequential_episodes(number, show)
    else:
        episodes = get_random_episodes(number, show)

    # Don't bother doing any of the below if there are no episodes
    if len(episodes) > 0:
        create_playlist(episodes)

        if autoplay == 'true':
            execute_json('Player.Open', {"item": {"playlistid": xbmc.PLAYLIST_VIDEO}})
            log('Playing playlist', level=xbmc.LOGDEBUG)
        else:
            execute_json('GUI.ActivateWindow', {'window': 'videoplaylist'})
            log('Showing playlist', level=xbmc.LOGDEBUG)


def create_menu(show_list):
    """Creates the 'folder' menu"""

    # TODO: Create 'entire library' list item
    # li = xbmcgui.ListItem('ALL TV SHOWS', iconImage='DefaultFolder.png')
    # xbmcplugin.addDirectoryItem(handle=addon_handle, url='show', listitem=li, isFolder=False)

    # 'Random show' entry
    if random_show_list_item == 'true':
        random_show = random.choice(show_list)
        url = sys.argv[0] + "?path=/play&show=%s" % (random_show['tvshowid'])
        li = xbmcgui.ListItem('- RANDOM SHOW -', iconImage='DefaultFolder.png')
        xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=li, isFolder=False)

    for s in show_list:
        url = sys.argv[0] + "?path=/play&show=%s" % (s['tvshowid'])
        show_artwork = get_show_details(s['tvshowid'], ['art'])['art']
        
        li = xbmcgui.ListItem(s['label'], iconImage=show_artwork.get('poster'),
                              thumbnailImage=show_artwork.get('poster'))
        li.setArt({'banner': show_artwork.get('banner'), 'fanart': show_artwork.get('fanart')})
        xbmcplugin.addDirectoryItem(handle=addon_handle, url=url, listitem=li, isFolder=False)
    
    xbmcplugin.endOfDirectory(addon_handle)

# The 'navigation' part
if path is None:
    create_menu(get_shows())

elif path[0] == '/play':
    showid = args.get('show', None)
    create_and_play(showid[0])